### ISSUE TITLE
Multiple States & Multiple Outputs

### ISSUE THREAD

**Issue by Nachoeigu** [langchain-ai/langchain-academy#9]
Hi,

After watching the module regarding States and Memory, I have two questions:

1) Imagine the following code, where as input we receive a query, and the output is the answer of that question. But the node needs to place part of its output in one state and the rest of its output in other state. How does langgraph handle this?

```python
from typing_extensions import TypedDict
from langgraph.graph import StateGraph, START, END
from helpers import analyzing_query

class OverallState(TypedDict):
    question: str
    answer: str

class ReasoningState(TypedDict):
    explanation: str

def node_1(state: OverallState): #It should return both states -> ReasoningState, OverallState
    print("---Node 1: analyze query and return output---")
    explanation, answer = analyzing_query(state['question'])
    return {
        #to OverallState it should place the answer in the answer key
        #to ReasoningState it should place the explanation in the explanation key
    }

builder = StateGraph(OverallState)
builder.add_node("node_1", node_1)

builder.add_edge(START, "node_1")
builder.add_edge("node_1", END)

graph = builder.compile()
```

2) Imagine another scenario, where the node instead of returning 2 updates of different states, the node needs to use TWO (or more) states as input in order to apply some logics. Something like the code below. How does LangGraph handle this?:
```python
class OverallState(TypedDict):
    question: str
    answer: str

class ReasoningState(TypedDict):
    explanation: str

class EvaluatorState(TypedDict):
    is_approved: bool

def evaluate_answer(overallstate: OverallState, reasoningstate: ReasoningState, evaluatorstate: EvaluatorState):
    evaluator_output = evaluating_llm_answer(overallstate['answer'], overallstate['question'])

    if evaluator_output is False: #This could mean that the answer is not correct
        evaluator_giving_correction = making_correction(reasoningstate['explanation'],overallstate['answer'], overallstate['question']) # This llm evaluates the explanation in order to correct the other llm in its reasoning
        return {
            #it returns the correct explanation, and the correct answer, placing each one in its corresponding state (it should return 2 states)
        }
    else:
        return {
            'is_approved': True #it should be over the evaluator state only. Keep the rest of states, without changes.
        }
```

Basically, the questions are regarding how LangGraph handle cases where a node, router or edge handle multiple states.

PD: It would be amazing if you can add a "Discussion" section so the students can communicate each other and improve the learning experience.

**Comment by robbiemu** [comment]
+1 about that discussions comment -- GitHub even has a discussions section that could be configured to have discussions for each module

**Comment by rlancemartin** [comment]
I like the discussion section idea! Discussing w/ @jessicaou.

**Comment by jessicaou** [comment]
Just added it to the platform. You can find it as a button for each module video.

**Comment by shiv248** [comment]
@robbiemu or join the slack [https://www.langchain.com/join-community](https://www.langchain.com/join-community) I believe you'll get faster responses here from the community in general.

**Comment by robbiemu** [comment]
> @Nachoeigu @robbiemu or join the slack https://www.langchain.com/join-community I believe you'll get more collaborative answers here from the community.

Thanks for the reminder about that. I was already joined but hadn't seen the hub, I joined for the runtime logs in the class workbooks :)

**Comment by rlancemartin** [comment]
Thanks! We are also working on a Chat-LangGraph QA interface that will aid this question-answering about LangGraph.

**Comment by rlancemartin** [comment]
> Imagine the following code, where as input we receive a query, and the output is the answer of that question. But the node needs to place part of its output in one state and the rest of its output in other state. How does langgraph handle this?

This is a great question, and we will add better conceptual docs on it. 

1/ **graph state** is the union of `OverallState` and `ReasoningState` 
2/ node **input schema** can be specified as you show, `state: OverallState`
3/ however, the node can write to any key in the **graph state** - e.g., `answer` and `explanation`

**Comment by rlancemartin** [comment]
> Hi,
> 
> After watching the module regarding States and Memory, I have two questions:
> 
> 1. Imagine the following code, where as input we receive a query, and the output is the answer of that question. But the node needs to place part of its output in one state and the rest of its output in other state. How does langgraph handle this?
> 
> ```python
> from typing_extensions import TypedDict
> from langgraph.graph import StateGraph, START, END
> from helpers import analyzing_query
> 
> class OverallState(TypedDict):
>     question: str
>     answer: str
> 
> class ReasoningState(TypedDict):
>     explanation: str
> 
> def node_1(state: OverallState): #It should return both states -> ReasoningState, OverallState
>     print("---Node 1: analyze query and return output---")
>     explanation, answer = analyzing_query(state['question'])
>     return {
>         #to OverallState it should place the answer in the answer key
>         #to ReasoningState it should place the explanation in the explanation key
>     }
> 
> builder = StateGraph(OverallState)
> builder.add_node("node_1", node_1)
> 
> builder.add_edge(START, "node_1")
> builder.add_edge("node_1", END)
> 
> graph = builder.compile()
> ```
> 
> 2. Imagine another scenario, where the node instead of returning 2 updates of different states, the node needs to use TWO (or more) states as input in order to apply some logics. Something like the code below. How does LangGraph handle this?:
> 
> ```python
> class OverallState(TypedDict):
>     question: str
>     answer: str
> 
> class ReasoningState(TypedDict):
>     explanation: str
> 
> class EvaluatorState(TypedDict):
>     is_approved: bool
> 
> def evaluate_answer(overallstate: OverallState, reasoningstate: ReasoningState, evaluatorstate: EvaluatorState):
>     evaluator_output = evaluating_llm_answer(overallstate['answer'], overallstate['question'])
> 
>     if evaluator_output is False: #This could mean that the answer is not correct
>         evaluator_giving_correction = making_correction(reasoningstate['explanation'],overallstate['answer'], overallstate['question']) # This llm evaluates the explanation in order to correct the other llm in its reasoning
>         return {
>             #it returns the correct explanation, and the correct answer, placing each one in its corresponding state (it should return 2 states)
>         }
>     else:
>         return {
>             'is_approved': True #it should be over the evaluator state only. Keep the rest of states, without changes.
>         }
> ```
> 
> Basically, the questions are regarding how LangGraph handle cases where a node, router or edge handle multiple states.
> 
> PD: It would be amazing if you can add a "Discussion" section so the students can communicate each other and improve the learning experience.

I added a new section to our docs to address this -- 
https://langchain-ai.github.io/langgraph/concepts/low_level/#multiple-schemas

Is anything unclear after reading it?

**Comment by Nachoeigu** [comment]
> > Hi,
> > After watching the module regarding States and Memory, I have two questions:
> > 
> > 1. Imagine the following code, where as input we receive a query, and the output is the answer of that question. But the node needs to place part of its output in one state and the rest of its output in other state. How does langgraph handle this?
> > 
> > ```python
> > from typing_extensions import TypedDict
> > from langgraph.graph import StateGraph, START, END
> > from helpers import analyzing_query
> > 
> > class OverallState(TypedDict):
> >     question: str
> >     answer: str
> > 
> > class ReasoningState(TypedDict):
> >     explanation: str
> > 
> > def node_1(state: OverallState): #It should return both states -> ReasoningState, OverallState
> >     print("---Node 1: analyze query and return output---")
> >     explanation, answer = analyzing_query(state['question'])
> >     return {
> >         #to OverallState it should place the answer in the answer key
> >         #to ReasoningState it should place the explanation in the explanation key
> >     }
> > 
> > builder = StateGraph(OverallState)
> > builder.add_node("node_1", node_1)
> > 
> > builder.add_edge(START, "node_1")
> > builder.add_edge("node_1", END)
> > 
> > graph = builder.compile()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. Imagine another scenario, where the node instead of returning 2 updates of different states, the node needs to use TWO (or more) states as input in order to apply some logics. Something like the code below. How does LangGraph handle this?:
> > 
> > ```python
> > class OverallState(TypedDict):
> >     question: str
> >     answer: str
> > 
> > class ReasoningState(TypedDict):
> >     explanation: str
> > 
> > class EvaluatorState(TypedDict):
> >     is_approved: bool
> > 
> > def evaluate_answer(overallstate: OverallState, reasoningstate: ReasoningState, evaluatorstate: EvaluatorState):
> >     evaluator_output = evaluating_llm_answer(overallstate['answer'], overallstate['question'])
> > 
> >     if evaluator_output is False: #This could mean that the answer is not correct
> >         evaluator_giving_correction = making_correction(reasoningstate['explanation'],overallstate['answer'], overallstate['question']) # This llm evaluates the explanation in order to correct the other llm in its reasoning
> >         return {
> >             #it returns the correct explanation, and the correct answer, placing each one in its corresponding state (it should return 2 states)
> >         }
> >     else:
> >         return {
> >             'is_approved': True #it should be over the evaluator state only. Keep the rest of states, without changes.
> >         }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Basically, the questions are regarding how LangGraph handle cases where a node, router or edge handle multiple states.
> > PD: It would be amazing if you can add a "Discussion" section so the students can communicate each other and improve the learning experience.
> 
> I added a new section to our docs to address this -- https://langchain-ai.github.io/langgraph/concepts/low_level/#multiple-schemas
> 
> Is anything unclear after reading it?

Yes, what about cases where a node writes over multiple schemas like one key in PrivateSchema and one key in overall state?

The same question for nodes that needs info as input of two schemas simultaneously (for example it needs as input values in PrivateSchema and in OverallSchema. Thanks @rlancemartin

**Comment by BharathRajM** [comment]
Hey,
We had the same need where a node needs to write to 2 different states, and also a node which can take inputs from 2 different states as @Nachoeigu mentioned at the top. 

This documentation (https://langchain-ai.github.io/langgraph/concepts/low_level/#multiple-schemas) takes into only single state into account. 

@Nachoeigu Were you able to find a way to implement for 2 different states with a single node? Appreciate the help
Thanks in advance

**Comment by ishanExtreme** [comment]
@Nachoeigu @BharathRajM 

I faced the similar issue faced here updating and reading from multiple state here's what I did to separate out these concerns

Case: 1

```python
from typing_extensions import TypedDict
from langgraph.graph import StateGraph, START, END
from helpers import analyzing_query

class GraphState(TypeDict):
    question: str
    answer: str

## internal/private state used by some/one nodes
class ReasoningState(TypedDict):
    explanation: str

class OverallState(GraphState, ReasoningState):
    pass



def node_1(state: GraphState) -> OverallState:
    print("---Node 1: analyze query and return output---")
    explanation, answer = analyzing_query(state['question'])
    return {
       "answer": answer,
        "explanation": explanation
    }

builder = StateGraph(GraphState)
builder.add_node("node_1", node_1)

builder.add_edge(START, "node_1")
builder.add_edge("node_1", END)

graph = builder.compile()
```


Case: 2

```python
class GraphState(TypedDict):
    question: str
    answer: str

## internal/private state used by some/one nodes
class ReasoningState(TypedDict):
    explanation: str

## internal/private state used by some/one nodes
class EvaluatorState(TypedDict):
    is_approved: bool

class OverallState(GraphState, ReasoningState, EvaluatorState):
    pass

def evaluate_answer(overallstate: OverallState) -> EvaluatorState:
    evaluator_output = evaluating_llm_answer(overallstate['answer'], overallstate['question'])

    if evaluator_output is False: #This could mean that the answer is not correct
        evaluator_giving_correction = making_correction(overallstate['explanation'],overallstate['answer'], overallstate['question']) # This llm evaluates the explanation in order to correct the other llm in its reasoning
        return {
          "is_approved": False
        }
    else:
        return {
            'is_approved': True
        }
```
